proxy/proxy.ts

```diff
@@ -0,0 +1,126 @@
+// @ts-ignore
+import isMartianIP from 'martian-cidr'
+import type http from 'node:http'
+import { createServer } from 'node:http'
+import { isIP } from 'node:net'
+import { styleText } from 'node:util'
+
+class BadRequestError extends Error {
+  constructor(message: string) {
+    super(message)
+    this.name = 'BadRequestError'
+  }
+}
+
+export const createProxyServer = (
+  config: {
+    fetchTimeout?: number
+    hostnameWhitelist?: string[]
+    isProduction?: boolean
+    productionDomainSuffix?: string
+    silentMode?: boolean
+  } = {}
+): http.Server => {
+  // Main toggle for production features
+  let isProduction = config.isProduction || false
+  // Silence the output. Used for testing
+  let silentMode = config.silentMode || false
+  // Allow request to certain ips like 'localhost'. Used for testing purposes
+  let hostnameWhitelist = config.hostnameWhitelist || []
+  // if isProduction, then only request from origins that match this param are allowed
+  let productionDomainSuffix = config.productionDomainSuffix || ''
+  let fetchTimeout = config.fetchTimeout || 2500
+
+  return createServer(async (req, res) => {
+    let url = decodeURIComponent(req.url!.slice(1))
+    let sent = false
+
+    try {
+      // Only http or https protocols are allowed
+      if (!url.startsWith('http://') && !url.startsWith('https://')) {
+        throw new BadRequestError('Only HTTP or HTTPS are supported')
+      }
+
+      // Other requests are typically used to modify the data, and we do not typically need them to load RSS
+      if (req.method !== 'GET') {
+        throw new BadRequestError('Only GET requests are allowed')
+      }
+
+      // In production mode we only allow request from production domain
+      if (isProduction) {
+        let origin = req.headers.origin
+        if (!origin?.endsWith(productionDomainSuffix)) {
+          throw new BadRequestError('Unauthorized Origin')
+        }
+      }
+
+      let requestUrl = new URL(url)
+      if (!hostnameWhitelist.includes(requestUrl.hostname)) {
+        // Do not allow requests to addresses that are reserved:
```

Comment:
```suggestion
        // Do not allow requests to addresses inside our cloud:
```

We should explain “why we did it” not “what we did”